src/Servers/Kestrel/Core/src/Internal/Http2/FlowControl/StreamOutputFlowControl.cs

```diff
@@ -83,7 +84,7 @@ public void Abort()
             // connection-level awaitable so the stream abort is observed immediately.
             // This could complete an awaitable still sitting in the connection-level awaitable queue,
             // but this is safe because completing it again will just no-op.
-            _currentConnectionLevelAwaitable?.Complete();
+            _currentConnectionLevelAwaitable?.TrySetResult(null);
```

Comment:
This seems dangerous. Is there any guarantee that the connection-level awaitable wasn't  dequeued already? I don't think there is.

I don't think TrySetResult is saving us either, because that's only checking `_core.GetStatus(_core.Version) == ValueTaskSourceStatus.Pending`. `_core.GetStatus(_core.Version)` is always going to validate, since we're not tracking the version independently, and it's very possible that the awaitable was already reused by the AvailabilityAwaitable getter meaning the that the status could be Pending, but pending for a different stream than the one that was aborted.

Tracking the version independently by storing it in AdvanceUpToAndWait and verifying it in Abort seems like the way to fix this, but there's always going to be the race with the _currentConnectionLevelAwaitable being completed naturally due to a connection level WINDOW_UPDATE right as the stream is aborted unless we find a way to synchronize those two things.

Reply:
Thinking about this a little more, completing an awaitable prematurely might not be the end of the world, since the continuation would make sure there is indeed enough availability for it to write, and if not, await again. I'd still rather we avoided ever completing awaitables prematurely if we have a good way of doing so though.

Reply:
Can TrySetResult be called here AFTER the awaitable has been dequeued?

A thing I worry about here is the awaitable being completed and dequeued, reused, and then being completed here.

Reply:
> Can TrySetResult be called here AFTER the awaitable has been dequeued?

Yes. That's what I am concerned about too. The easiest fix is probably to change ManualResetValueTaskSource.TrySetResult to take a version token. In AdvanceUpToAndWait, when we decide to store a _currentConnectionLevelAwaitable, we could also store the original version in a _currentConnectionLevelAwaitableVersion field.